New syllabi included for Spring 2023 and Summer 2023

A278PlusPi opened the pull request:
This includes:
- **CSCI 39535**: UI/UX Design (Spring 2023 | Prof. Melissa Lynch)
- **CSCI 499**: Advanced Applications: A Capstone for Majors (Spring 2023 | Prof. Artjom Plaunov)
- **CSCI 340**: Operating Systems (Summer 2023 | Prof. Stanley Wine)

joshnatis commented:
Wow! This is great, very interesting to see these two new syllabi (and it's good to have an update for Wine's).

Btw, I've written the stuff below for educational purposes, if you're pressed for time / don't care, you don't have to read any of it, just put in this suggested change and we'll be good:

**TLDR**: rename this syllabus as follows:
```diff
- CS340_wine_syllabus_sum23.pdf
+ CS340_wine_syllabus_f23.pdf
```

## One necessary change before merging

There is one thing that we'll have to change before merging, and it's due to a silly limitation of the website (which we should fix). The 340 syllabus you uploaded is from the summer semester, and you've (reasonably) named it `CS340_wine_syllabus_sum23.pdf`, but that will not work due to (faulty) assumptions made by our code processing these syllabi files -- it expects either `f` for fall or `s` for spring.

This is the code (in [build/syllabus.py](https://github.com/Hunter-Open-Source-Club/syllabi/blob/354c571eac2991c4222b7678a640e63bf5e183e4/build/syllabus.py#L17-L19)) that we use to break down the file name into its parts (course number, professor name, semester):
```python
  1 def parse(syllabus_name):
  2         syllabus = Syllabus()
  3         syllabus.file_name = syllabus_name
  4 
  5         tokens = syllabus_name.split('_')
  6 
  7         try:
  8                 syllabus.course    = tokens[0]
  9                 syllabus.professor = tokens[-3].capitalize()
 10                 syllabus.semester  = tokens[-1].replace('.pdf','').upper()
 11 
 12                 season_map      = { 'F': 'Fall', 'S': 'Spring' }
 13                 syllabus.season = season_map[syllabus.semester[0]]
 14                 syllabus.year   = int(syllabus.semester[1:])
 15         except:
 16                 utils.log('Invalid syllabus filename:', syllabus_name)
 17                 return None
 18 
 19         return syllabus
```

So given a syllabus name like `CS340_wine_syllabus_sum23.pdf`, we extract the semester, e.g. `sum23`, and then we try to break it up into the `season` and the `year`. Lines 12-14 are the culprit which cause this to fail for `sum`, but work for `f` and `s`.

Because we don't have / receive many few syllabi from the Winter and Summer terms, this hasn't proved too big of an annoyance to fix. Here is one I can find, where I simply labeled the file name as `f` even though it was a summer course: http://syllabi.hunterosc.org/syllabi/CS/STAT_213/STAT213_jalayer_syllabus_f19.pdf, hence motivating my recommendation to you in this case.

We should fix this in the future, I've created an issue here: https://github.com/Hunter-Open-Source-Club/syllabi/issues/38

## Testing

It's good practice to test your change to make sure everything works as expected before merging. If I had merged this without noticing the problem with the file name, this would have made http://syllabi.hunterosc.org go down until we noticed the issue, found the bug, and then corrected it. As a maintainer it's my job to keep a close eye and test things myself, and as a contributor it's good ettiquette for you to test things on your end and let me know (or even prove via screenshots or something) in your PR description that you tested stuff and it looks good. But in both of those roles people get lazy and are busy :)

We added this section in the README for testing locally: https://github.com/Hunter-Open-Source-Club/syllabi#building-the-website-locally. There are no dependencies for building the website aside from a Unix shell (like bash) and python3, with the hope that you'll be able to run it on your end with no problem. (You can even run it on ENIAC if you don't have a computer with a Unix-like OS at your disposal). 

With your changes, when I try to build the `CS` department's syllabus bank page, I run into this error:
```sh
$ python3 build/build.py CS
Invalid syllabus filename: CS340_wine_syllabus_sum23.pdf
Traceback (most recent call last):
  File "build/build.py", line 92, in <module>
    main()
  File "build/build.py", line 24, in main
    courses    = generate_course_html()
  File "build/build.py", line 29, in generate_course_html
    discover_syllabi_files(course_map)
  File "build/build.py", line 45, in discover_syllabi_files
    'files': utils.sort_chronologically(syllabi_files)
  File "/Users/josh/comp/src/syllabi/build/utils.py", line 31, in sort_chronologically
    return sorted(syllabi_files, key=functools.cmp_to_key(compare))
  File "/Users/josh/comp/src/syllabi/build/utils.py", line 26, in compare
    if s.season == 'Fall':
AttributeError: 'NoneType' object has no attribute 'season'
```

So if you tried this on your system you would know something is wrong. Upon closer inspection, there is an error message in there signifying something wrong with the syllabus name:
> Invalid syllabus filename: CS340_wine_syllabus_sum23.pdf

But then we have some further errors happening downstream due to not being able to parse the syllabus name earlier. In an ideal world we would simply skip syllabi with invalid names rather than failing to build. I've created an issue for this as well: https://github.com/Hunter-Open-Source-Club/syllabi/issues/39

After renaming the syllabus name, everything builds smoothly for me:
```sh
$ python3 build/build.py CS
$
$ # No error message, we're good!
```
<img width="691" alt="image" src="https://github.com/Hunter-Open-Source-Club/syllabi/assets/31445542/36bc0339-7061-4435-a868-c698b104d29c">

The more complex your change is, the more important it is for you to prove to the maintainer that your change is working by explaining how you tested. Even better than that, the maintainer can set up some tests which automatically run against each change, proving whether it works or not. The key word for this is CI (continuous integration). If I had that set up here, then GitHub would have shown us a big red X telling us that something is wrong. Would you look at that, another improvement we can make -- I've created an issue for it: https://github.com/Hunter-Open-Source-Club/syllabi/issues/40. As our industry matures, we're trying to make our software more reliable by relying on tools to validate things for us rather than just counting on our brains and hoping to never make a mistake.

## Questions for you

I have two quick questions:

1. Was there anything unclear about the instructions in the README? Is there anything you read which confused you or you feel can be improved?
    + For one, it said nothing about what to do when you have a summer or winter syllabus, and there's no precedent for it when you look at existing filenames, so you were forced to guess `sum23`. That's not very nice.
2. What was giving you trouble while creating the directories? Maybe I can help for next time?

Thanks again for these contributions, and I apologize for my super long reply lol, hopefully there is at least one helpful thing in there

A278PlusPi commented:
### Updates:

First, I have added Tojeira's CSCI 335 syllabi as well as Weiss' CSCI 340 syllabi (the difference for the latter is that it now has the Zoom link and the QR code)

Second, I have renamed Prof. Wine's syllabus from `Sum23` to `F23` for the parser to accept it given the information it has. When I read your comment about file name parsing (at Line 12 and 13), it currently only takes a single character. This is understandable for syllabi for Spring and Fall terms (and even Winter terms - in that case, you would use `W` for Winter). However, for syllabi in the summer, we may have to consider making changes to the parser. If changes aren't ideal, we can use `U` or `M` for Summer syllabi, but it won't be pretty.

Third, I initially had trouble making file directories as the way I did it is making the files in GitHub's website, and I have to put some file there in the folder (otherwise the folder is deleted). I then realized that cloning the GitHub repo makes things easier for me so I didn't have to be making unnecessary commits every single time.

And finally, with all the changes I have made, I have decided to following the build instructions, and it generated HTML files, and it works just fine on my end. Thank you so much for the long comment you have made in this pull request. I hope that your comment will be useful not only to me, but to everyone else that contribute to open source projects like this and needs some advice and best practices. I hope that more people will be willing to upload more syllabi for computer science and beyond! 🚀
